python/ray/data/tests/test_streaming_executor.py

```diff
@@ -0,0 +1,180 @@
+
+from typing import List, Any
+
+import ray
+from ray.data.context import DatasetContext
+from ray.data._internal.execution.interfaces import ExecutionOptions, RefBundle
+from ray.data._internal.execution.streaming_executor import StreamingExecutor
+from ray.data._internal.execution.streaming_executor_state import (
+    OpState,
+    build_streaming_topology,
+    process_completed_tasks,
+    select_operator_to_run,
+)
+from ray.data._internal.execution.operators.all_to_all_operator import AllToAllOperator
+from ray.data._internal.execution.operators.map_operator import MapOperator
+from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+from ray.data._internal.execution.util import make_ref_bundles
+
+
+@ray.remote
+def sleep():
+    time.sleep(999)
+
+
+def make_transform(block_fn):
+    def map_fn(block_iter):
+        for block in block_iter:
+            yield block_fn(block)
+
+    return map_fn
+
+
+def ref_bundles_to_list(bundles: List[RefBundle]) -> List[List[Any]]:
+    output = []
+    for bundle in bundles:
+        for block, _ in bundle.blocks:
+            output.append(ray.get(block))
+    return output
+
+
+def test_build_streaming_topology():
+    inputs = make_ref_bundles([[x] for x in range(20)])
+    o1 = InputDataBuffer(inputs)
+    o2 = MapOperator(make_transform(lambda block: [b * -1 for b in block]), o1)
+    o3 = MapOperator(make_transform(lambda block: [b * 2 for b in block]), o2)
+    topo, _ = build_streaming_topology(o3)
+    assert len(topo) == 3, topo
+    assert o1 in topo, topo
+    assert not topo[o1].inqueues, topo
+    assert topo[o1].outqueue == topo[o2].inqueues[0], topo
+    assert topo[o2].outqueue == topo[o3].inqueues[0], topo
+
+
+def test_process_completed_tasks():
+    inputs = make_ref_bundles([[x] for x in range(20)])
+    o1 = InputDataBuffer(inputs)
+    o2 = MapOperator(make_transform(lambda block: [b * -1 for b in block]), o1)
+    topo, _ = build_streaming_topology(o2)
+
+    # Test processing output bundles.
+    assert len(topo[o1].outqueue) == 0, topo
+    process_completed_tasks(topo)
+    assert len(topo[o1].outqueue) == 20, topo
+
+    # Test processing completed work items.
+    sleep_ref = sleep.remote()
+    done_ref = ray.put("done")
+    o2.get_work_refs = MagicMock(return_value=[sleep_ref, done_ref])
+    o2.notify_work_completed = MagicMock()
+    o2.inputs_done = MagicMock()
+    process_completed_tasks(topo)
+    o2.notify_work_completed.assert_called_once_with(done_ref)
+    o2.inputs_done.assert_not_called()
+
+    # Test input finalization.
+    o2.get_work_refs = MagicMock(return_value=[done_ref])
+    o2.notify_work_completed = MagicMock()
+    o2.inputs_done = MagicMock()
+    o1.completed = MagicMock(return_value=True)
+    topo[o1].outqueue.clear()
+    process_completed_tasks(topo)
+    o2.notify_work_completed.assert_called_once_with(done_ref)
+    o2.inputs_done.assert_called_once()
+
+
+def test_select_operator_to_run():
```

Comment:
If we're relying on topological ordering being the tie-break, could we have a test for that?

Reply:
I think the tie-breaking order isn't that important right now.